Sprint 20

Hajime4life opened the pull request:
Sprint 20

Hajime4life committed:
- свёрстан компонент для отображения историй

Hajime4life committed:
Добавления:
- Полностью переработан компонент историй
- Разделена вся логика работы историй
- Добавлен прогресс бар
- Теперь истории перелистываются
- * Добавлена анимация после переноса истории в "просмотренные"
- Переходы между историями реализованы с помощью TabView

NaR0RoW commented on TravelSchedule/ViewModel/SettingsViewModel.swift:
```diff
@@ -0,0 +1,13 @@
+import SwiftUI
+
+class SettingsViewModel: ObservableObject {
```
Стоит добавить модификатор `final`, чтобы вызовы методов работали быстрее.
Но это уже со звездочкой, хотя на собеседованиях спрашивают. Можно почитать про [диспачеризацию тут](https://habr.com/ru/articles/673400/)

NaR0RoW commented on TravelSchedule/ViewModel/SettingsViewModel.swift:
```diff
@@ -0,0 +1,13 @@
+class SettingsViewModel: ObservableObject {
+    @Published var isDark: Bool {
+        didSet {
+            UserDefaults.standard.set(isDark, forKey: "isDarkMode")
```
Лучше добавить сервис для работы с `UserDefaults`. Например что-то такое:
``` swift
final class UserDefaultsService {
    static let shared = UserDefaultsService()
    private let defaults = UserDefaults.standard

    private init() {}

    private enum Key {
        static let someKey = "someKey"
    }

    var isSomeKeyCompleted: Bool {
        get { defaults.bool(forKey: Key.someKey) }
        set { defaults.set(newValue, forKey: Key.someKey) }
    }
}
```

Также рекомендую посмотреть в сторону [AppStorage](https://developer.apple.com/documentation/swiftui/appstorage)

NaR0RoW commented on TravelSchedule/Views/PrivacyPolicyView.swift:
```diff
@@ -0,0 +1,63 @@
+    }
+}
+
+class WKWebViewController: UIViewController {
```
Стоит отметить как `final`. Перепроверь пожалуйста остальные классы в коде. Если класс не наследуется, то лучше отметить его как `final` для ускорения работы компилятора

Hajime4life replied:
- [ ] ok

NaR0RoW commented on TravelSchedule/Views/PrivacyPolicyView.swift:
```diff
@@ -0,0 +1,63 @@
+
+class WKWebViewController: UIViewController {
+    var url: URL?
+    private var webView: WKWebView!
```
Лучше не использовать `force` операции

NaR0RoW commented on TravelSchedule/Views/PrivacyPolicyView.swift:
```diff
@@ -0,0 +1,63 @@
+        }
+        .onDisappear() {
+            navigation.showTabBar = .visible
+        }
```
Круглые скобки можно опустить:
``` swift
.onAppear {
   navigation.showTabBar = .hidden
}
.onDisappear {
   navigation.showTabBar = .visible
}
```

NaR0RoW commented on TravelSchedule/Views/PrivacyPolicyView.swift:
```diff
@@ -0,0 +1,63 @@
+    }
+}
+
+struct WebView: UIViewControllerRepresentable {
```
Структуру можно сделать приватной

NaR0RoW commented on TravelSchedule/Views/Stories/Helper/TimerConfiguration.swift:
```diff
@@ -0,0 +1,29 @@
+        self.progressPerTick = 1.0 / CGFloat(storiesCount) / secondsPerStory * timerTickInternal
+    }
+    
+    func progress(for storyIndex: Int) -> CGFloat {
+        return min(CGFloat(storyIndex) / CGFloat(storiesCount), 1)
+    }
+
+    func index(for progress: CGFloat) -> Int {
+        return min(Int(progress * CGFloat(storiesCount)), storiesCount - 1)
+    }
+
+    func nextProgress(progress: CGFloat) -> CGFloat {
+        return min(progress + progressPerTick, 1)
+    }
```
В однострочных выражениях `return` можно опускать

NaR0RoW commented on TravelSchedule/Views/Stories/StoriesListView.swift:
```diff
@@ -0,0 +1,51 @@
+struct StoriesListView: View {
+    @StateObject private var storiesViewModel = StoryViewModel()
+    @State private var showStoryInFullscreen: Bool = false
+    @State private var selectedStoryIdx: Int? = nil
```
`= nil` можно опустить

NaR0RoW commented on TravelSchedule/Views/Stories/StoriesListView.swift:
```diff
@@ -0,0 +1,51 @@
+                    startIndex: selectedStoryIdx,
+                    storiesCount: storiesViewModel.storiesCount,
+                )
+                .environmentObject(storiesViewModel)
```
Форматирование кода немного поехало, можно выделить это место и воспользоваться автоматическим форматированием в Xcode `control + I`

NaR0RoW commented on TravelSchedule/Views/Stories/StoriesView.swift:
```diff
@@ -0,0 +1,83 @@
+    
+    @Binding var show: Bool
+    @State var currentStoryIndex: Int
+    @State var currentProgress: CGFloat
```
`State property wrapper` в `SwiftUI` должен быть приватным, чтобы гарантировать, что изменение значений происходит только в теле представления. Для иных целей стоит использовать `@Binding`.
Так что поле стоит отметить как `private`

NaR0RoW commented on TravelSchedule/Views/Stories/StoryView.swift:
```diff
@@ -0,0 +1,50 @@
+            .resizable()
+            .aspectRatio(contentMode: .fill)
+            .frame(maxWidth: UIScreen.main.bounds.width)
+            .edgesIgnoringSafeArea(.top)
```
Лучше ` .ignoresSafeArea()` использовать. `.edgesIgnoringSafeArea()` больше не актуальна и не будет поддерживаться в будущих версиях iOS

Hajime4life committed:
Изменения:
- добавлен final на непереиспользуемые классы
- форс операции больше не используются в webview
- webview был заменен статическим текстом
- истории больше не отображаются в .sheet, теперь это fullscreen

Hajime4life commented:
Я видел что в ТЗ написано истории должны быть на fullscreen потом зашел в макет фигмы и ни одного экрана где нарисован fullscreen экран для историй. Наоборот в последних темах объяснялось как правильно завернуть углы с одной стороны, вот я задолбался и завернул углы снизу на истории как в макете, а оказалось это всё ненужно и всё было в 10 раз проще) а за это еще отказали :(

Hajime4life commented:
И как я только не пытался реализовать темную тему на webview - не смог)

NaR0RoW commented:
> Я видел что в ТЗ написано истории должны быть на fullscreen потом зашел в макет фигмы и ни одного экрана где нарисован fullscreen экран для историй. Наоборот в последних темах объяснялось как правильно завернуть углы с одной стороны, вот я задолбался и завернул углы снизу на истории как в макете, а оказалось это всё ненужно и всё было в 10 раз проще) а за это еще отказали :(

Может быть не совсем очевидно, но это fullScreen. Просто мы в нем "не игнорируем" safeArea, из-за чего View растягивается не на весь экран:

<img width="328" height="636" alt="Screenshot 2025-08-13 at 14 17 46" src="https://github.com/user-attachments/assets/5fef609d-97f3-4b99-881b-c69f9fa290c1" />

Если бы это был sheet, то был бы виден экран "под". Как пример:

<img width="376" height="714" alt="Screenshot 2025-08-13 at 14 19 13" src="https://github.com/user-attachments/assets/8b4a7bb9-eee8-4673-a07b-ee9b8aeb6a89" />

fullScreen:

<img width="427" height="715" alt="Screenshot 2025-08-13 at 14 19 39" src="https://github.com/user-attachments/assets/81a41df7-4517-434f-bb91-7675ccc1d41e" />

NaR0RoW commented:
> Я видел что в ТЗ написано истории должны быть на fullscreen потом зашел в макет фигмы и ни одного экрана где нарисован fullscreen экран для историй. Наоборот в последних темах объяснялось как правильно завернуть углы с одной стороны, вот я задолбался и завернул углы снизу на истории как в макете, а оказалось это всё ненужно и всё было в 10 раз проще) а за это еще отказали :(

Насчет углов, я так понимаю, что там что-то такое используется:
``` swift
struct RoundedCorners: Shape {
    let radius: CGFloat
    let corners: UIRectCorner

    func path(in rect: CGRect) -> Path {
        let path = UIBezierPath(
            roundedRect: rect,
            byRoundingCorners: corners,
            cornerRadii: CGSize(width: radius, height: radius)
        )
        return Path(path.cgPath)
    }
}
```
В целом, это стандартная практика на всех проектах

NaR0RoW commented:
> И как я только не пытался реализовать темную тему на webview - не смог)

Обычно [используют](https://stackoverflow.com/questions/69830561/wkwebview-dark-and-light-mode-with-dynamic-url-in-swift-5) CSS
